optax/_src/utils.py

```diff
@@ -333,6 +333,178 @@ def _value_and_grad(
+    """Returns True if any element in the sequence is a container or array."""
+
+    iterable = seq.values() if isinstance(seq, dict) else seq
+
+    for v in iterable:
+        if _is_container(v):
+            return True
+        # Any JAX / NumPy / array-like object exposes .dtype
+        if hasattr(v, "dtype"):
+            return True
+    return False
+
+
+def _is_bool_scalar(x):
+    """Checks if `x` represents a single boolean value."""
+    
+    # Python bool
+    if isinstance(x, bool):
+        return True
+    # 0-D JAX/Numpy array of dtype bool_
+    if isinstance(x, jnp.ndarray) and x.shape == () and x.dtype == jnp.bool_:
+        return True
+    return False
+
+
+def _broadcast_mask(params: Any, mask: Any, *, _root: bool = True) -> Any:
+    """Recursively expands a boolean mask potentially matching a prefix of params.
+
+    Args:
+      params: The PyTree defining the target structure and leaves.
+      mask: A boolean scalar (Python `bool` or JAX `jnp.bool_`), or a PyTree
+        (dict, list, tuple) of booleans representing a full or prefix mask.
+      _root: Internal flag indicating if currently at the top level of recursion.
+
+    Returns:
+      A boolean PyTree with the exact same structure as `params`, where mask
+      values are propagated according to prefix rules and defaults. Missing
+      branches or leaves default to False.
+    """
+    if _is_bool_scalar(mask):
+        flag = bool(mask)
+
+        # At the root level, a scalar mask applies to all leaves of params.
+        if _root:
+            return tree_util.tree_map(lambda _: flag, params)
+
+        if _is_container(params):
+            # If params is a container with actual subtrees or arrays, broadcast
+            # the flag down. This prevents masking trivial structures like
+            # lists of simple Python numbers if they appear as parameters.
+            if _has_pytree_or_array(params):
+                return tree_util.tree_map(lambda _: flag, params)
+            else:
+                return tree_util.tree_map(lambda _: False, params)
+
+        return flag
+
+    if isinstance(params, dict):
+        m_dict = mask if isinstance(mask, dict) else {}
+        return {k: _broadcast_mask(p, m_dict.get(k, False), _root=False)
+                for k, p in params.items()}
+
+    if isinstance(params, list):
+        mask_seq = list(mask) if isinstance(mask, (list, tuple)) else []
+        mask_seq = (mask_seq + [False] * len(params))[: len(params)]
+        return [_broadcast_mask(p, m, _root=False)
+                for p, m in zip(params, mask_seq)]
+
+    if isinstance(params, tuple):
+        mask_seq = list(mask) if isinstance(mask, (list, tuple)) else []
+        mask_seq = (mask_seq + [False] * len(params))[: len(params)]
+        return tuple(_broadcast_mask(p, m, _root=False)
+                     for p, m in zip(params, mask_seq))
+
+    return False
+
+
+def freeze(mask: base.PyTree) -> base.GradientTransformation:
```

Comment:
let's rely on jax's prefix tree behavior for jax.tree.map and require explicity prefix/leaf False - for more predictable behavior